numba/tests/test_lists.py

```diff
@@ -865,5 +865,270 @@ def test_reflect_clean(self):
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_pop_loc(self):
+        def pyfunc():
+            con = []
+            for i in range(1000):
+                con.append(np.arange(i + 1))
+            while len(con) > 2:
+                con.pop(1)
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_del_range(self):
+        def pyfunc():
+            con = []
+            for i in range(20):
+                con.append(np.arange(i + 1))
+            del con[3:10]
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assert_list_element_precise_equal(
+            expect=expect, got=got
+            )
+
+    def test_list_of_list(self):
+        def pyfunc():
+            con = []
+            for i in range(10):
+                con.append([0] * i)
+            return con
+
+        cfunc = jit(nopython=True)(pyfunc)
+        expect = pyfunc()
+        got = cfunc()
+
+        self.assertEqual(expect, got)
+
+    def test_list_of_list_reflected(self):
+        def pyfunc(l1, l2):
+            l1.append(l2)
+            l1[-1].append(123)
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l1 = [[0, 1], [2, 3]]
+        l2 = [4, 5]
+        expect = list(l1), list(l2)
+        got = list(l1), list(l2)
+        pyfunc(*expect)
+        cfunc(*got)
+        self.assertEqual(expect, got)
+
+    def test_heterogenous_list(self):
+        def pyfunc(x):
+            return x[1]
+
+        l1 = [[np.zeros(i) for i in range(5)], [np.ones(i) for i in range(5)]]
+        l2 = [[np.zeros(i) for i in range(5)], [np.ones(i)+1j for i in range(5)]]
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l1_got = cfunc(l1)
+        self.assertPreciseEqual(pyfunc(l1), l1_got)
+        l2_got = cfunc(l2)
+        self.assertPreciseEqual(pyfunc(l2), l2_got)
+
+    def test_heterogeneous_list_error(self):
+        def pyfunc(x):
+            return x[1]
+
+        cfunc = jit(nopython=True)(pyfunc)
+        l3 = [[np.zeros(i) for i in range(5)], [(1,),]]
+        l4 = [[1], [{1}]]
+        l5 = [[1], [{'a':1}]]
+
+        # TODO:
+        # working_cases = [l1, l2]
+
+        # error_cases
+        with self.assertRaises(TypeError) as raises:
+            cfunc(l3)
+        self.assertIn("can't unbox array", str(raises.exception))
+
+        if utils.PY2:
+            int_error = "long() argument must be a string"
+        else:
+            int_error = "int() argument must be a string"
+        with self.assertRaises(TypeError) as raises:
+            cfunc(l4)
+        self.assertIn(int_error, str(raises.exception))
+        self.assertIn("not 'set'", str(raises.exception))
```

Comment:
This may or may not be for this PR, but I think the docs need some information adding about how the type of a list is determined (and promotion rules). With the error that appears when using `l4`, I do not think it is clear to users what went wrong. I also wonder if it is possible to catch the `TypeError` when coming from `setitem` in a list context and try and explain a bit about what the issue is during the exception handling (same goes for trying to unbox heterogeneous lists)? Perhaps a `Loc` instance could go in the function descriptor to help with pointing to exactly where things went wrong (debatable as to whether baking in line number info is useful?)?

Reply:
I agree that we need to make better error messages in the unboxing of list.  I vote for doing that in a separate issue.